src/d3d12/nvapi_d3d12_device.cpp

```diff
@@ -31,14 +32,40 @@ namespace dxvk {
         if (cubinDevice == nullptr)
             return false;
 
-        return SUCCEEDED(cubinDevice->CreateCubinComputeShaderWithName(cubinData, cubinSize, blockX, blockY, blockZ, shaderName, reinterpret_cast<D3D12_CUBIN_DATA_HANDLE**>(pShader)));
+        return CreateCubinComputeShaderEx(device, cubinData, cubinSize, blockX, blockY, blockZ, 0 /* smemSize */, shaderName, pShader);
+    }
+
+    bool NvapiD3d12Device::CreateCubinComputeShaderEx(ID3D12Device* device, const void* cubinData, NvU32 cubinSize, NvU32 blockX, NvU32 blockY, NvU32 blockZ, NvU32 smemSize, const char* shaderName, NVDX_ObjectHandle* pShader) {
+        auto cubinDevice = GetCubinDevice(device);
+        if (cubinDevice == nullptr)
+            return false;
+
+        Com<ID3D12DeviceExt> deviceExt;
+        if (FAILED(cubinDevice->QueryInterface(IID_PPV_ARGS(&deviceExt))))
+            return false;
+
+        bool status = SUCCEEDED(deviceExt->CreateCubinComputeShaderWithName(cubinData, cubinSize, blockX, blockY, blockZ, shaderName, reinterpret_cast<D3D12_CUBIN_DATA_HANDLE**>(pShader)));
+
+        if (status == true) {
+            std::scoped_lock lock(m_CubinSmemMutex);
+            m_cubinSmemMap.emplace(*pShader, smemSize);
+        }
+
+        return status;
```

Comment:
That raises the scope that the `std::scoped_lock` exists in, which means that we'd be holding the lock when calling into vkd3d-proton. This means that if multiple threads call `NvAPI_D3D12_CreateCubinShader*()` at the same time, they'll execute serially rather than in parallel.

Reply:
Doesn't the `scoped_lock` in above example protect only the statements after its construction until the function returns? The interface call should be outside its scope and hence executed without the lock being held because it's constructed only after the interface calls returns so the only code protected by it is the `emplace` call and the evaluation of returned `true` constant.

Reply:
A sample to illustrate the point:

```cpp
#include <format>
#include <iostream>
#include <mutex>
#include <pthread.h>
#include <thread>

void proc()
{
    static std::mutex m;
    auto tid = pthread_self();

    std::cout << std::format("Hello from thread {}!\n", tid); // not protected, this could be the interface call

    // { not needed
    std::scoped_lock<std::mutex> lock{m};

    std::this_thread::sleep_for(std::chrono::seconds(2)); // protected, this could be the emplace
    std::cout << std::format("Goodbye from thread {}!\n", tid);
    // } not needed
}

int main()
{
    std::thread a{proc};
    std::thread b{proc};
    std::thread c{proc};
    std::thread d{proc};

    a.join();
    b.join();
    c.join();
    d.join();
}
```

Reply:
C vs C++ semantics, whoops :) Fixed in latest push